inference-engine/include/ie_blob.h

```diff
@@ -136,6 +156,14 @@ class INFERENCE_ENGINE_API_CLASS(Blob) {
         return tensorDesc;
     }
 
+    /**
+     * @brief Returns ROI data for the blob when it was specified. For the blob without ROI the pointer will contain
+     * nothing.
+     */
+    virtual const std::unique_ptr<ROIData>& getROI() const noexcept {
```

Comment:
it's `const ROIData` indeed, no change to the data is assumed

Reply:
So return value should be `const ROIData&`, right?

Reply:
RV `const std::unique_ptr<const ROIData>&`

Reply:
Why so, you said not change to the data is assumed, but user still can change ROIData

Reply:
in what way?

Reply:
`operator *` for `const std::unique_ptr<T>` returns non constant reference `T&` (https://en.cppreference.com/w/cpp/memory/unique_ptr/operator*).

So `const std::unique_ptr<T>` allows to modify the object it points to.

Reply:
>RV const std::unique_ptr<const ROIData>&

Reply:
Let me show a snippet of code:

```cpp
// Example program
#include <iostream>
#include <string>
#include <memory>

struct Foo {
    std::unique_ptr<int> bar;
    Foo() {
        bar = std::make_unique<int>(4);
    }
    const std::unique_ptr<int>& foo() const {
        return bar;
    }
};

int main()
{
  Foo foo;
  const auto& bar = foo.foo();
  std::cout << *bar << std::endl; // 4
  *bar = 5; // compiles
  std::cout << *bar << std::endl; // 5
}

```

Reply:
> `RV const std::unique_ptr<const ROIData>&`

Have you seen the reply few lines above?

Reply:
Sorry, I didn't notice change. My question is the same why you need to specify `std::unique_ptr` here at all? It doesn't add any meaning here, only confusion: `const ROIData&` is simpler and has the same effect

Reply:
The pointer contains null when ROI is absent.

Reply:
Agreed, so it means we should have `const ROIData*` ?

Reply:
>`RV const std::unique_ptr<const ROIData>&`

The object pointer is more convenient to handle by the user. Consider `if` check

Reply:
I don't think so, you still can do the same check with usual raw pointer. `std::unique_ptr` is confusing here, because purpose of the class is introduce specific ownership strategy - it is known to have *value* semantic, so using it over reference is ill formed interface.

Reply:
It's neither confusing, not leaking.
The user is told explicitly that he shouldn't try to store the pointer somewhere since it won't live longer than the object.

Reply:
I still disagree,

`The user is told explicitly that he shouldn't try to store the pointer somewhere` - you don't tell that - you don't give him a pointer - you give hime reference to pointer.

`it won't live longer than the object` - that's true for both options raw pointer and reference to `std::unique_ptr`

I think we need invite 3rd side to give a comment. @vinograd47 what do you think?

Reply:
Ok, here is more involved explanation:
>The user is told explicitly that he shouldn't try to _store the pointer_ somewhere

The raw pointer (*) is meant.
The user, being given raw pointer, might save it for later usage, and use it even after the data get destroyed.
Being told about the scope with `unique_ptr`, there is far lesser chance for storing the pointer.

Does it make sense?

Reply:
Not much as I see, the user, being given reference to `std::unique_ptr`, might save it for later usage and use it even after the data get destroyed.

Reply:
Consider the difference between the context within the client code where: - reference can be **initialized** - and the raw pointer be **copied** and passed over. Does not it reduce the chances? Storing the reference to the object takes much thoughts on the structure of the code in contrast to the raw ptr. Client would not intentionally break his code having noticed the `unique_ptr`.

Reply:
I'd like to summarise my opinion here:

`const std::unique_ptr<const ROIData>&` is too complex to me without enough profit. It doesn't speak about ownership, but has `std::unique_ptr` that's confusing. It may reduce chances user will save (or may not) return value and get dangling reference - there we're trying to say to user - don't do it if you don't want to get dangling reference.

My preferable alternative is `const ROIData*` - it is much simpler, prevents user from modifications, allows to check on `nullptr` as criteria of non-roi object. It speaks about ownership explicitly and in common way (raw pointer does not own and extend object lifetime). There is an objective it's easier to save it somewhere in comparison with reference, so can be considered error prone. IMHO, we already made it clear that returned value could be dangling (true for first option as well), so it is up to user how to deal with - it can be saved and work as expected at the same time (if user manages to use it only in correct context).

At the same time, I see there is no other opinion which I treat as agreement with both options (or just ignorance of topic), so if it is, I think it's ok to leave code as is. Especially if my concerns was took into consideration.

Feel free to resolve.

Reply:
Here we deal with the optional field, which may be present or not. Therefore returning a reference or a copy is not an option.

Sharing raw pointer is the simplest option - but it does not declare the scope of the pointer valid state (it can be guessed although).
Sharing `unique_ptr` has the complexer signature - but has the profit of not leaving the current scope without significant efforts from the user.
Sharing `shared_ptr` is the safest alternative, but it comes for the price of an additional reference counter.

From these options I took the middle.